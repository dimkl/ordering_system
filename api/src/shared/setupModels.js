const { Model } = require('objection');
const knex = require('./knex');

module.exports = () => { Model.knex(knex) }