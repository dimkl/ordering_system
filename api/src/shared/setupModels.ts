import { Model } from "objection";
import { knex } from "./knex";

export default () => {
  Model.knex(knex);
  return knex;
};
