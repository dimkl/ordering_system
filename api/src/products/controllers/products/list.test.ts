/**
 * @integration-test true
 * @data-factory true
 */
import type { Knex } from "knex";

import setupModels from "../../../shared/setupModels";

describe("GET /products", () => {
  let knex: Knex;
  beforeAll(() => (knex = setupModels()));
  beforeEach(() => knex.raw("truncate users, products, ingredients cascade;"));
  afterAll(() => knex.destroy());

  it("list products", async () => {
    const product = await DataFactory.createProduct();
    await DataFactory.createProductIngredient({}, product);

    const response = await request.get(`/${apiVersion}/products`).set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body.length).toBe(1);
    expect(response.body[0]).toMatchObject({
      id: product.id,
      created_at: expect.any(String),
      updated_at: expect.any(String),
      title: product.title,
      description: product.description
    });
    expect(response.body[0].ingredients).toBeUndefined();
  });

  it("list products with included ingredients", async () => {
    const product = await DataFactory.createProduct();
    await DataFactory.createProductIngredient({}, product);

    const response = await request
      .get(`/${apiVersion}/products?include=ingredients`)
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body.length).toBe(1);
    expect(response.body[0]).toMatchObject({
      id: product.id,
      created_at: expect.any(String),
      updated_at: expect.any(String),
      title: product.title,
      description: product.description
    });
    expect(response.body[0].ingredients.length).toBe(1);
  });

  it("throws 400 for not existing included param", async () => {
    const response = await request
      .get(`/${apiVersion}/products?include=variations`)
      .set("Accept", "application/json");

    expect(response.status).toEqual(400);
    expect(response.body).toMatchSnapshot();
  });
});
