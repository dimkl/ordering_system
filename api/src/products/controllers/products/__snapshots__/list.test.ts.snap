// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /products throws 400 for not existing included param 1`] = `
{
  "message": "Provided resource is not supported! Update the \`include\` query param with the correct value(s).",
}
`;
